### PR TITLE
Add option to tweak inductor stride settings for user-defined triton kernels

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135581
* __->__ #135530

Previously, Inductor was allowed to modify the stride/storage_offset
(layout) for inputs to user-defined triton kernels. This can cause
silent incorrectness because most triton kernels are written for a
specific striding pattern (usually contiguous).

This PR adds a config to allow the user to choose Inductor's behavior on
this. The options are:
- "flexible_layout" (default): Inductor can modify the layout for inputs
  to user-defined triton kernels as much as it wants.
- "needs_fixed_stride_order": Inductor must preserve the stride order
  (when compared to tracing) for inputs to user-defined triton kernels.

This matches our handling for custom operators. In the future, we'll
want a "needs_exact_strides" option (this is the safest option).

Test Plan:
- new test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -1303,6 +1323,25 @@ def debug(msg: str) -> None:
...
                 self.propagate_mutation(n, old_args, old_kwargs, args, kwargs)  # type: ignore[possibly-undefined]
+            elif (
+                n.op == "call_function"
+                and n.target is torch.ops.higher_order.triton_kernel_wrapper_mutation
+                and config.triton_kernel_default_layout_constraint != "flexible_layout"
+            ):
+                debug("user_defined_triton_kernel_layout_constraints")
+                if (
+                    config.triton_kernel_default_layout_constraint
+                    == "needs_fixed_stride_order"
```
Comment (FindHao) [inline torch/_inductor/graph.py:1334]: Will strings `needs_fixed_stride_order` and `flexible_layout` be used in other places? If so, can we avoid hardcoding?

**Review by FindHao — APPROVED** [review]
LGTM

**Review by oulgen — APPROVED** [review]

File: torch/_inductor/graph.py (hunk 1 above)
Comment (zou3519) [reply torch/_inductor/graph.py:1333]: I don't think so, I'll turn it into an enum if I end up needing it more. I wasn't sure if enums were valid Inductor config options (there's a limited set of types it accepts).

**Review by eellison — COMMENTED** [review]
Do we still want stride order as the api now that we support matching exact strides ?

**Comment by zou3519** [comment]
> Do we still want stride order as the api now that we support matching exact strides ?

We haven't put the exact strides logic in yet. The API should be {flexible_layout, match_stride_order, match_exact_strides}, and the ultimate default we want is match_exact_strides. We're moving there but "match_stride_order" is a good middle ground.

**Pull request closed without merging**